Objects/longobject.c

```diff
@@ -6775,3 +6776,115 @@ int PyLong_AsUInt64(PyObject *obj, uint64_t *value)
 {
     LONG_TO_UINT(obj, value, "C uint64_t");
 }
+
+
+static const PyLongLayout PyLong_LAYOUT = {
+    .bits_per_digit = PyLong_SHIFT,
+    .digits_order = -1,  // least significant first
+    .digit_endianness = PY_LITTLE_ENDIAN ? -1 : 1,
+    .digit_size = sizeof(digit),
+};
+
+
+const PyLongLayout*
+PyLong_GetNativeLayout(void)
+{
+    return &PyLong_LAYOUT;
+}
+
+
+int
+PyLong_Export(PyObject *obj, PyLongExport *export_long)
+{
+    if (!PyLong_Check(obj)) {
+        PyErr_Format(PyExc_TypeError, "expect int, got %T", obj);
+        return -1;
+    }
+
+    // Fast-path: try to convert to a int64_t
+    int overflow;
+#if SIZEOF_LONG == 8
+    long value = PyLong_AsLongAndOverflow(obj, &overflow);
+#else
+    // Windows has 32-bit long, so use 64-bit long long instead
+    long long value = PyLong_AsLongLongAndOverflow(obj, &overflow);
+#endif
+    Py_BUILD_ASSERT(sizeof(value) == sizeof(int64_t));
```

Comment:
This ought to be equally fast and portable without any platform assumptions.

```c
int64_t value;
int r = PyLong_AsNativeBytes(obj, &value, sizeof(value), Py_ASNATIVEBYTES_NATIVE_ENDIAN);
assert (!(r < 0 && PyErr_Occurred());
if (r <= sizeof(value)) {
    // fast path success
    export_long->value = value;
}
```

Reply:
For an extra fast path, replace `sizeof(value)` with the size of a compact value (IIRC, `sizeof(Py_ssize_t)-1` is the best you'll get) and then it's not going to do the bit conversion for values close to `2**64`.

Reply:
(And yes, I believe we should replace the implementation of `PyLong_AsCompilerSpecificType()` functions with `PyLong_AsNativeBytes(o, &retval, sizeof(retval), ...)` calls, apart from laziness and the _guaranteed_ compatibility. Can't break it if you never change it!)

Reply:
This code is performance critical and we spend significant time to find the most efficient function calls, it's about a few nanoseconds: https://peps.python.org/pep-0757/#export-pylong-export-with-gmpy2

PyLong_AsLongAndOverflow() and PyLong_AsLongLongAndOverflow() are simple and efficient. PyLong_AsNativeBytes() is less complex if I recall correctly.

Reply:
Another reason for using this API was "no-error" contract.  Maybe we can specify that in docs as a CPython implementation detail?  IIUC, we are free to change such things in new releases without deprecation period.

Note also, that gmpy2 benchmarks measure not just CPython side, but the whole conversion path (int->gmpy2.mpz in this case).

Reply:
They're no more simple or efficient than `AsNativeBytes`, and with the options I specified and the fact we know the value is `PyLong_Check`ed then there are no exceptions possible from `AsNativeBytes` (you'll notice in my alternate code I kept the `assert` for exceptions).

If we're actually going for greatest speed, then we should be checking for a compact value here anyway, and exposing values too big for a CV but still within 64-bits as digits, rather than converting them at all.

Reply:
As I said, we tested several variants.  Here some benchmarks:
https://github.com/python/cpython/pull/121339#issuecomment-2357588919
FYI, "current implementation" essentially refers to https://github.com/python/cpython/pull/121339/commits/d70a1214a675cab71cc87b8935cd2798c81c9f96 (inlined version of PyLong_AsInt64).

@zooba, do you think we need to repeat tests?

Reply:
This code is so close to the PyLong implementation, that I think we should use `_PyLong_IsCompact()` + `_PyLong_CompactValue()` to be sure that it matches the specification.